Importer/CsvImporter.php

```diff
@@ -23,88 +27,241 @@ class CsvImporter extends AbstractImporter
+    {
+        $this->headers = $headers;
+    }
+
+    /**
+     * @param string $delimiter
+     */
+    public function setDelimiter($delimiter)
+    {
+        $this->delimiter = $delimiter;
+    }
+
+    /**
+     * @param InterpreterErrorHandlerInterface $lineErrorHandler
+     */
+    public function registerInterpreterErrorHandler(InterpreterErrorHandlerInterface $lineErrorHandler)
+    {
+        $this->interpreter->registerErrorHandler($lineErrorHandler);
+    }
+
+    /**
+     * @param ImporterErrorHandlerInterface $fileErrorHandler
+     */
+    public function registerImporterErrorHandler(ImporterErrorHandlerInterface $fileErrorHandler)
+    {
+        $this->importerErrorHandlers[] = $fileErrorHandler;
+    }
+
+    /**
+     * @return ConfigurationInterface
+     */
+    public function getConfiguration()
+    {
+        return $this->configuration;
     }
 
     /**
      * @param string $filename
-     * @param bool   $hasHeaders
      * @param bool   $dryRun
-     *
-     * @return array|null|object[]
-     * @throws DatabaseException
      */
-    public function importFile($filename, $hasHeaders = true, $dryRun = false)
+    public function importFile($filename, $dryRun = false)
     {
+        if (!file_exists($filename)) {
+            $this->log->addConfigurationError('File not found');
+
+            return;
+        }
+
         $csv = file_get_contents($filename);
+
+        $this->import($csv, $dryRun);
+    }
+
+    /**
+     * @param string $csv
+     * @param bool   $dryRun
+     */
+    public function import($csv, $dryRun = false)
+    {
         try {
-            return $this->import($csv, $hasHeaders, $dryRun);
-        } catch (DatabaseException $exception) {
-            $exception->setDataFile($filename);
-            throw $exception;
+            if ($this->hasHeaders() && !$this->checkHeadersAreValid($csv)) {
+                return;
+            }
+
+            $this->log->setCsv($csv);
+
+            $parsedData = $this->parser->parse($csv, $this->hasHeaders(), $this->delimiter);
+
+            $this->importData($parsedData, $this->hasHeaders(), !$dryRun);
+        } catch (\Throwable $throwable) {
+            $message = sprintf(
+                '%s: %s (uncaught exception) at %s line %d while running import.',
+                get_class($throwable),
+                $throwable->getMessage(),
+                $throwable->getFile(),
+                $throwable->getLine()
+            );
+            $this->log->addCriticalError($message);
         }
     }
 
     /**
-     * @param array $csv
-     * @param bool  $hasHeaders
-     * @param bool  $dryRun
-     *
-     * @return array|null
-     * @throws DatabaseException
+     * @param array $parsedData
+     * @param bool  $dataIsAssociativeArray
+     * @param bool  $flush
      */
-    public function import($csv, $hasHeaders = true, $dryRun = false)
+    protected function importData(array $parsedData, $dataIsAssociativeArray, $flush = true)
```

Comment:
`$flush`?
Technically it's fine, and it's a private method (well, protected, but still not accessible from outside) but maybe it's better to keep standard of using `$dryRun` logic, especially that it's opposite to `$flush`.